docs/agents.md

```diff
@@ -1,128 +1,233 @@
+  signal: AbortSignal.timeout(60000),
+  execution: {
+    maxIterations: 20,
+    maxRetriesPerStep: 3,
+    totalMaxRetries: 10,
+  },
+});
+```
 
-Hence, the agent in each iteration produces one of the following outputs.
+#### `observe(callback: (emitter: Emitter) => void): void`
 
-For the sake of simplicity, imagine that the input prompt is "What is the current weather in Las Vegas?"
+Subscribes to agent events for monitoring and debugging.
 
-First iteration:
+```ts
+agent.observe((emitter) => {
+  // Listen for complete updates
+  emitter.on("update", ({ data, update, meta }) => {
+    console.log(`Complete Update: ${update.key} = ${update.value}`);
+  });
 
-```
-thought: I need to retrieve the current weather in Las Vegas. I can use the OpenMeteo function to get the current weather forecast for a location.
-tool_name: OpenMeteo
-tool_input: {"location": {"name": "Las Vegas"}, "start_date": "2024-10-17", "end_date": "2024-10-17", "temperature_unit": "celsius"}
+  // Listen for partial updates (streaming)
+  emitter.on("partialUpdate", ({ data, update, meta }) => {
+    console.log(`Partial Update: ${update.key} = ${update.value}`);
+  });
+
+  // Listen for tool execution
+  emitter.on("toolStart", ({ tool, input }) => {
+    console.log(`Tool Started: ${tool.name}`);
+  });
+});
 ```
 
-> [!NOTE]
->
-> Agent emitted 3 complete updates in the following order (`thought`, `tool_name`, `tool_input`) and tons of partial updates in the same order.
-> Partial update means that new tokens are being added to the iteration. Updates are always in strict order: You first get many partial updates for thought, followed by a final update for thought (that means no final updates are coming for a given key).
+## Events
 
-Second iteration:
+Agent emits various events through its DevTools.Emitter:
 
-```
-thought: I have the current weather in Las Vegas in Celsius.
-final_answer: The current weather in Las Vegas is 20.5°C with an apparent temperature of 18.3°C.
-```
+| Event           | Description                | Payload                   |
+| --------------- | -------------------------- | ------------------------- |
+| `update`        | Complete update for a step | `{ data, update, meta }`  |
+| `partialUpdate` | Streaming update           | `{ data, update, meta }`  |
+| `toolStart`     | Tool execution started     | `{ tool, input }`         |
+| `toolEnd`       | Tool execution completed   | `{ tool, result }`        |
+| `error`         | Error occurred             | `{ error, context }`      |
+| `retry`         | Retry attempt              | `{ attempt, maxRetries }` |
+| `success`       | Successful completion      | `{ result }`              |
 
-For more complex tasks, the agent may do way more iterations.
+## Implementation Example
 
-In the following example, we will transform the knowledge gained into code.
+Here's an example of implementing an simple agent base in Bee Agent class:
```

Comment:
This is not the implementation of a base class; it is just a showcase of its usage.